primitives/src/merkle_tree/field_based_mht/ramht/poseidon.rs

```diff
@@ -0,0 +1,342 @@
+
+
+                let (input_vec, output_vec) =
+                    self.array_nodes[self.initial_pos_subarray[i]..self.final_pos_subarray[i + 1]].split_at_mut(self.final_pos_subarray[i] - self.initial_pos_subarray[i]);
+
+                let new_pos_parent = self.new_elem_pos_subarray[i + 1] + num_groups_leaves;
+
+                PoseidonBatchHash::<F,P>::batch_evaluate_in_place(&mut input_vec[(self.processed_pos_subarray[i] - self.initial_pos_subarray[i])..(last_pos_to_process - self.initial_pos_subarray[i])],
+                                           &mut output_vec[(self.new_elem_pos_subarray[i + 1] - self.initial_pos_subarray[i + 1])..(new_pos_parent - self.initial_pos_subarray[i + 1])]);
+
+                self.new_elem_pos_subarray[i + 1] += num_groups_leaves;
+                self.new_elem_pos[i + 1] += num_groups_leaves;
+                self.processed_pos_subarray[i] += num_groups_leaves * self.rate;
+                self.processed_pos[i] += num_groups_leaves * self.rate;
+
+                if self.new_elem_pos_subarray[i] == self.final_pos_subarray[i] {
+                    self.new_elem_pos_subarray[i] = self.initial_pos_subarray[i];
+                }
+                if self.processed_pos_subarray[i] == self.final_pos_subarray[i] {
+                    self.processed_pos_subarray[i] = self.initial_pos_subarray[i];
+                }
+            }
+        }
+    }
+}
+
+impl<F: PrimeField + MulShort, P: Clone + PoseidonParameters<Fr=F>> RandomAccessMerkleTree
+for PoseidonRandomAccessMerkleTree<F, P> {
+
+    type Data = F;
+    type MerklePath = PoseidonRAMTPath<F>;
+
+    fn init(num_leaves: usize) -> Self {
+        // change it to the rate of the hash function
+        let rate = P::R;
+        let processing_step = num_leaves;
+
+        let cpus = rayon::current_num_threads();
+        let mut chunk_size = processing_step / (cpus * rate);
+        let mut processing_block_size = chunk_size * cpus * rate;
+        if processing_step < cpus * rate {
+            chunk_size = processing_step / rate;
+            if chunk_size == 0 {
+                chunk_size = 1;
+            }
+            processing_block_size = chunk_size * rate;
+        }
+        // the processing block is calculated to be a multiple of the number of cpus and the rate
+        // if the processing step is not an exact multiple of the cpu and the rate then it is rounded down
+
+        let last_level_size = num_leaves.next_power_of_two();
+        // last_level_size is the minimum power of two that contains the leaves
+        // if the number of leaves is smaller than a power of two then they will be filled by zero at the end.
+        let mut size = last_level_size;
+
+        assert!(last_level_size >= processing_block_size, "The number of leaves should be bigger than the processing chunk size.");
+
+        if processing_block_size > last_level_size {
+            // if the processing block is greater than the number of leaves, then
```

Comment:
It seems to me that you are setting always processing_block_size = processing_step = num_leaves (except when chunk_size = 0). Is this correct ?